feat: duplicate navigation related APIs to `contents.navigationHistory`

alicelovescake opened the pull request:
#### Description of Change

This PR refactors and duplicates navigation related APIs to `contents.navigationHistory`. Follow up to previous PR https://github.com/electron/electron/pull/41577  where a new navigationHistory property is created on web contents. 

This PR duplicates and exposes the following API for consistency and clarity:
- `canGoBack`
- `goBack`
- `canGoForward`
- `goForward`
- `canGoToOffset`
- `goToOffset`
- `clear`

Added related tests and documentation.

For backwards compatibility, the above APIs can still be directly accessed from webContents. 

<!--
Thank you for your Pull Request. Please provide a description above and review
the requirements below.

Contributors guide: https://github.com/electron/electron/blob/main/CONTRIBUTING.md
-->

#### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [x] relevant documentation, tutorials, templates and examples are changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/main/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/main/README.md#examples).

#### Release Notes

Notes: Added the following existing navigation related APIs to `webcontents.navigationHistory`: `canGoBack`, `goBack`, `canGoForward`, `goForward`,  `canGoToOffset`, `goToOffset`, `clear`

codebytere reviewed (commented):
Codewise this looks great to me - the only outstanding question here is what (if any) timeline we'd want to consider targeting for deprecation of the old methods. Our options I would say are:

1) Merge this as is and support both for the medium-term future, deprecating the old approach in a handful of versions and removing a few after that
2) Deprecate now to indicate that developers should look to use the new code where possible but leave in for a good long time, no specific removal date
3) Deprecate and remove in the next few targeted cycles

My personal approach would probably be #2 but I think any are reasonably justifiable. Thoughts (cc @MarshallOfSound)

samuelmaddock commented:
I'd be in favor of adding deprecation warnings now (2) as to give developers and library authors the most time to migrate.

alicelovescake commented:
Thanks for the feedback! @codebytere @samuelmaddock 

I updated the PR with `_Deprecated_` in the docs to let developers know to use the new api.

In the future, when we are ready to officially deprecate, I can:
- update docs/breaking-changes.md
- move the api to internal-electron.d.ts


Let me know if there is anything else I can do for deprecation warnings for this PR!

codebytere reviewed (commented):
@alicelovescake sorry for the delay here - in addition to adding deprecation notes to docs, we'll also want to leverage our internal `deprecate` module to notify developers at run-time as well:

```js
deprecate.warnOnce('webContents.canGoToOffset', 'webContents.navigationHistory.canGoToOffset');
```

alicelovescake commented:
> @alicelovescake sorry for the delay here - in addition to adding deprecation notes to docs, we'll also want to leverage our internal `deprecate` module to notify developers at run-time as well:
> 
> ```js
> deprecate.warnOnce('webContents.canGoToOffset', 'webContents.navigationHistory.canGoToOffset');
> ```

Thanks for the code pointer! Added the warnings by:
- [x] Moved and renamed existing navigation APIs to internal. (i.e webContents.canGoBack => webContents._canGoBack)
- [x] Added javascript wrappers to emit the deprecation warning: I.e
```
const canGoBackDeprecated = deprecate.warnOnce('webContents.canGoBack', 'webContents.navigationHistory.canGoBack');
WebContents.prototype.canGoBack = function () {
  canGoBackDeprecated();
  return this._canGoBack();
};
```
- [x] Tested that warning works correctly: image below
<img width="947" alt="image" src="https://github.com/electron/electron/assets/66543449/662b12b3-8e1e-4bda-a8e8-92a990e74bf3">

codebytere reviewed (approved):
API LGTM

samuelmaddock commented on docs/api/web-contents.md:
```diff
@@ -1126,44 +1126,60 @@ Reloads the current web page.
 Navigates browser to the specified absolute web page index.
 
-#### `contents.goToOffset(offset)`
+**Deprecated:** Should use the new [`contents.navigationHistory.goToIndex`](navigation-history.md#navigationhistorygotoindexindex) API.
```
```suggestion
**Deprecated:** Should use the new [`contents.navigationHistory.goToIndex`](navigation-history.md#navigationhistorygotoindex) API.
```

samuelmaddock commented on docs/api/web-contents.md:
```diff
@@ -1126,44 +1126,60 @@ Reloads the current web page.
 
 Navigates to the specified offset from the "current entry".
 
+**Deprecated:** Should use the new [`contents.navigationHistory.goToOffset`](navigation-history.md#navigationhistorygotooffsetoffset) API.
```
```suggestion
**Deprecated:** Should use the new [`contents.navigationHistory.goToOffset`](navigation-history.md#navigationhistorygotooffset) API.
```

samuelmaddock commented on docs/api/web-contents.md:
```diff
@@ -1126,44 +1126,60 @@ Reloads the current web page.
 Returns `boolean` - Whether the web page can go to `offset`.
 
-#### `contents.clearHistory()`
+**Deprecated:** Should use the new [`contents.navigationHistory.canGoToOffset`](navigation-history.md#navigationhistorycangotooffsetoffset) API.
```
```suggestion
**Deprecated:** Should use the new [`contents.navigationHistory.canGoToOffset`](navigation-history.md#navigationhistorycangotooffset) API.
```

alicelovescake replied:
The links are actually correct because the duplicated word represents the argument! I clicked on and double checked all the links to make sure they work!

Thanks for the detailed review!

samuelmaddock commented on docs/api/web-contents.md:
```diff
@@ -1126,44 +1126,60 @@ Reloads the current web page.
 Clears the navigation history.
 
-#### `contents.goBack()`
+**Deprecated:** Should use the new [`contents.navigationHistory.clearHistory`](navigation-history.md#navigationhistoryclear) API.
```
```suggestion
**Deprecated:** Should use the new [`contents.navigationHistory.clear`](navigation-history.md#navigationhistoryclear) API.
```

samuelmaddock reviewed (approved):
API LGTM

erickzhao commented on docs/api/navigation-history.md:
```diff
@@ -9,6 +9,24 @@ Each navigation entry corresponds to a specific page. The indexing system follow
+
+* `offset` Integer
+
+Returns `boolean` - Whether the web page can go to `offset`.
```
```suggestion
Returns `boolean` - Whether the web page can go to the specified `offset` from the current entry.
```

I liked the wording in the older API better because it was clearer about where the offset was from.

erickzhao reviewed (approved):
docs: lgtm
